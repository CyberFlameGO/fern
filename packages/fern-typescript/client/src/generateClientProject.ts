import { IntermediateRepresentation } from "@fern-api/api";
import {
    DependencyManager,
    GeneratedProjectSrcInfo,
    generateTypeScriptProject,
    getOrCreateDirectory,
} from "@fern-typescript/commons";
import { generateEncoderFiles } from "@fern-typescript/encoders";
import { HelperManager } from "@fern-typescript/helper-manager";
import { generateModelFiles } from "@fern-typescript/model";
import { Volume } from "memfs/lib/volume";
import { Directory } from "ts-morph";
import { generateHttpService } from "./http/generateHttpService";
import { generateWebSocketChannel } from "./websocket/generateWebSocketChannel";

export async function generateClientProject({
    intermediateRepresentation,
    helperManager,
    packageName,
    packageVersion,
    volume,
}: {
    intermediateRepresentation: IntermediateRepresentation;
    helperManager: HelperManager;
    packageName: string;
    packageVersion: string;
    volume: Volume;
}): Promise<void> {
    await generateTypeScriptProject({
        volume,
        packageName,
        packageVersion,
        generateSrc: (directory) => generateClientFiles({ intermediateRepresentation, helperManager, directory }),
    });
}

async function generateClientFiles({
    intermediateRepresentation,
    helperManager,
    directory,
}: {
    intermediateRepresentation: IntermediateRepresentation;
    helperManager: HelperManager;
    directory: Directory;
}): Promise<GeneratedProjectSrcInfo> {
    const dependencyManager = new DependencyManager();

    const modelContext = generateModelFiles({
        modelDirectory: directory.createDirectory("model"),
        intermediateRepresentation,
        dependencyManager,
    });

    const encodersDirectory = getOrCreateDirectory(directory, "encoders");

    const servicesDirectory = getOrCreateDirectory(directory, "services");
    for (const service of intermediateRepresentation.services.http) {
        await generateHttpService({
            service,
            servicesDirectory,
            modelContext,
            encodersDirectory,
            helperManager,
            dependencyManager,
        });
    }

    for (const channel of intermediateRepresentation.services.websocket) {
        generateWebSocketChannel({
            channel,
            servicesDirectory,
            modelContext,
            encodersDirectory,
            helperManager,
            dependencyManager,
        });
    }

    await generateEncoderFiles({
        encodersDirectory,
        intermediateRepresentation,
        modelContext,
        servicesDirectory,
        helperManager,
    });

    return {
        dependencies: dependencyManager.getDependencies(),
    };
}
