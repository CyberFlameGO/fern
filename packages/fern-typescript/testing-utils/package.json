{
  "name": "@fern-typescript/testing-utils",
  "version": "0.0.0",
  "repository": {
    "type": "git",
    "url": "git@github.com:fern-api/fern.git",
    "directory": "packages/fern-typescript/testing-utils"
  },
  "files": [
    "lib"
  ],
  "type": "module",
  "source": "src/index.ts",
  "module": "src/index.ts",
  "main": "src/index.ts",
  "sideEffects": false,
  "scripts": {
    "clean": "rm -rf ./lib && tsc --build --clean",
    "compile": "tsc --build",
    "test": "yarn run compile && jest --passWithNoTests",
    "lint:eslint": "eslint --max-warnings 0 . --ignore-path=../../../.eslintignore",
    "lint:eslint:fix": "eslint --max-warnings 0 . --ignore-path=../../../.eslintignore --fix",
    "format": "prettier --write --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "format:check": "prettier --check --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "depcheck": "depcheck"
  },
  "dependencies": {
    "@fern-api/ir-generator": "workspace:*",
    "@fern-api/workspace-loader": "workspace:*",
    "@fern-fern/ir-model": "0.0.49",
    "@fern-typescript/commons": "workspace:*",
    "execa": "^5.1.1",
    "is-ci": "^3.0.1",
    "memfs": "^3.4.1",
    "typescript": "4.6.4"
  },
  "devDependencies": {
    "@babel/core": "^7.18.9",
    "@babel/preset-env": "^7.18.9",
    "@babel/preset-typescript": "^7.18.6",
    "@types/is-ci": "^3.0.0",
    "@types/jest": "^28.1.6",
    "@types/node": "^18.6.1",
    "depcheck": "^1.4.3",
    "eslint": "^8.20.0",
    "jest": "^28.1.3",
    "prettier": "^2.7.1",
    "ts-jest": "^28.0.7"
  }
}
