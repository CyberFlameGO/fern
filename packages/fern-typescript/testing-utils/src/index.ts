export { itFernETE } from "./itFernETE";
export { installAndCompileGeneratedProjects } from "./runEteTest";
