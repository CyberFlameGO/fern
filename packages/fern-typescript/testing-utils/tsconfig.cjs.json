{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/cjs", "rootDir": "src", "module": "CommonJS" },
  "include": ["./src"],
  "references": [
    { "path": "../../cli/tsconfig.cjs.json" },
    { "path": "../../ir/ir-generator/tsconfig.cjs.json" },
    { "path": "../../workspace-parser/tsconfig.cjs.json" },
    { "path": "../commons/tsconfig.cjs.json" }
  ]
}
