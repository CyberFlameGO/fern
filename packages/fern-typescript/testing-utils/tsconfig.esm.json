{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/esm", "rootDir": "src", "module": "esnext" },
  "include": ["./src"],
  "references": [
    { "path": "../../cli/tsconfig.esm.json" },
    { "path": "../../compiler/compiler/tsconfig.esm.json" },
    { "path": "../commons/tsconfig.esm.json" }
  ]
}
