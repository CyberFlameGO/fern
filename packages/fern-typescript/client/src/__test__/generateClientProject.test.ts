import { itFernETE } from "@fern-typescript/testing-utils";
import path from "path";
import { generateClientProject } from "../generateClientProject";

const FIXTURES_DIR = path.join(__dirname, "fixtures");
const FIXTURES = ["posts", "no-errors", "chat", "auth"];

describe("generateClientProject", () => {
    for (const fixture of FIXTURES) {
        itFernETE(fixture, {
            testFile: __filename,
            pathToFixture: path.join(FIXTURES_DIR, fixture),
            generateFiles: async ({ volume, intermediateRepresentation }) => {
                await generateClientProject({
                    packageName: fixture,
                    packageVersion: "0.0.0",
                    volume,
                    intermediateRepresentation,
                });
            },
        });
    }
});
