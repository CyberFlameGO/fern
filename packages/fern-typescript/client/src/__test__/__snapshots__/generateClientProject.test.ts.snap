// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateClientProject chat app 1`] = `
Object {
  "/package.json": "{
    \\"name\\": \\"chat\\",
    \\"version\\": \\"0.0.0\\",
    \\"files\\": [
        \\"lib\\"
    ],
    \\"main\\": \\"./lib/cjs/index.js\\",
    \\"types\\": \\"./lib/cjs/index.d.ts\\",
    \\"exports\\": {
        \\".\\": {
            \\"require\\": \\"./lib/cjs/index.js\\",
            \\"default\\": \\"./lib/esm/index.js\\"
        }
    },
    \\"sideEffects\\": false,
    \\"scripts\\": {
        \\"build\\": \\"run-p build:esm build:cjs\\",
        \\"build:esm\\": \\"tsc --project tsconfig.esm.json && echo '{ \\\\\\"type\\\\\\": \\\\\\"module\\\\\\" }' > lib/esm/package.json\\",
        \\"build:cjs\\": \\"tsc --project tsconfig.json && echo '{ \\\\\\"type\\\\\\": \\\\\\"commonjs\\\\\\" }' > lib/cjs/package.json\\"
    },
    \\"dependencies\\": {
        \\"uuid\\": \\"^8.3.2\\"
    },
    \\"devDependencies\\": {
        \\"@types/uuid\\": \\"^8.3.4\\",
        \\"@types/node\\": \\"^17.0.33\\",
        \\"npm-run-all\\": \\"^4.1.5\\",
        \\"typescript\\": \\"^4.6.4\\"
    }
}",
  "/src/model/chat/errors/GenericMessage.ts": "export type GenericMessage = string & {
    __GenericMessage: void;
};

export const GenericMessage = {
    of: (value: string): GenericMessage => value as GenericMessage
};
",
  "/src/model/chat/errors/index.ts": "export * from \\"./GenericMessage\\";
",
  "/src/model/chat/index.ts": "export * from \\"./types\\";
export * from \\"./errors\\";
export * from \\"./service-types\\";
",
  "/src/model/chat/service-types/ChatService/_SendChatErrorBody.ts": "import { chat } from \\"../../..\\";
import * as uuid from \\"uuid\\";

export type _SendChatErrorBody =
    | _SendChatErrorBody.GenericMessage;

export declare namespace _SendChatErrorBody {
    interface GenericMessage {
        _error: \\"GenericMessage\\";
        _errorInstanceId: string;
        GenericMessage: chat.GenericMessage;
    }

    export interface _Visitor<Result> {
        GenericMessage: (value: chat.GenericMessage) => Result;
        _unknown: () => Result;
    }
}

export const _SendChatErrorBody = {
    GenericMessage: (value: chat.GenericMessage): _SendChatErrorBody.GenericMessage => ({
        GenericMessage: value,
        _error: \\"GenericMessage\\",
        _errorInstanceId: uuid.v4()
    }),

    _visit: <Result>(value: _SendChatErrorBody, visitor: _SendChatErrorBody._Visitor<Result>): Result => {
        switch (value._error) {
            case \\"GenericMessage\\": return visitor.GenericMessage(value.GenericMessage);
            default: return visitor._unknown();
        }
    },

    _types: (): _SendChatErrorBody[\\"_error\\"][] => [\\"GenericMessage\\"],
} as const;
",
  "/src/model/chat/service-types/ChatService/_SendChatRequest.ts": "import * as model from \\"../../..\\";

export interface _SendChatRequest {
    id: string;
    operation: \\"sendChat\\";
    body: model.chat._SendChatRequestBody;
}
",
  "/src/model/chat/service-types/ChatService/_SendChatRequestBody.ts": "export interface _SendChatRequestBody {
    title: string;
    content: string;
}
",
  "/src/model/chat/service-types/ChatService/_SendChatResponse.ts": "import * as model from \\"../../..\\";

export type _SendChatResponse = _SendChatResponse.SuccessResponse | _SendChatResponse.ErrorResponse;

namespace _SendChatResponse {
    export interface SuccessResponse {
        ok: true;
        id: string;
        replyTo: string;
        body: model.chat.ChatMessage;
    }

    export interface ErrorResponse {
        ok: false;
        id: string;
        replyTo: string;
        error: model.chat._SendChatErrorBody;
    }
}
",
  "/src/model/chat/service-types/ChatService/index.ts": "export * from \\"./_SendChatRequestBody\\";
export * from \\"./_SendChatRequest\\";
export * from \\"./_SendChatErrorBody\\";
export * from \\"./_SendChatResponse\\";
",
  "/src/model/chat/service-types/index.ts": "export * from \\"./ChatService\\";
",
  "/src/model/chat/types/ChatMessage.ts": "import * as model from \\"../..\\";

export interface ChatMessage {
    id: model.chat.ChatMessageId;
    title: string;
    content: string;
}
",
  "/src/model/chat/types/ChatMessageId.ts": "export type ChatMessageId = string & {
    __ChatMessageId: void;
};

export const ChatMessageId = {
    of: (value: string): ChatMessageId => value as ChatMessageId
};
",
  "/src/model/chat/types/index.ts": "export * from \\"./ChatMessageId\\";
export * from \\"./ChatMessage\\";
",
  "/src/model/index.ts": "export * as chat from \\"./chat\\";
",
  "/src/services/chat/ChatService.ts": "import path from \\"path\\";
import * as model from \\"../../model\\";
import * as uuid from \\"uuid\\";

export declare namespace ChatService {
    interface Args {
        origin: string;
    }

    type ServerMessage = model.chat._SendChatResponse;
}

export interface ChatService {
    sendChat(request: model.chat._SendChatRequestBody): Promise<model.chat._SendChatResponse>;
}

export class ChatService implements ChatService {
    private socket: Promise<WebSocket>;
    private callbacks: Record<string, (response: any) => void> = {};

    constructor(args: ChatService.Args) {
        this.socket = new Promise(resolve => {
            const socket = new WebSocket(path.join(args.origin, \\"/chat\\"));
            socket.onopen = () => {
                resolve(socket);
            };
            socket.onmessage = this._onMessage;
        });
    }

    public async sendChat(request: model.chat._SendChatRequestBody): Promise<model.chat._SendChatResponse> {
        const socket = await this.socket;

        return new Promise(resolve => {
            const message: model.chat._SendChatRequest = {
                id: uuid.v4(),
                operation: \\"sendChat\\",
                body: request
            };
            this.callbacks[message.id] = resolve;
            socket.send(JSON.stringify(message));
        });
    }

    public async _disconnect(): Promise<void> {
        const socket = await this.socket;
        return new Promise(resolve => {
            socket.onclose = () => {
                resolve();
            };
            socket.close();
        });
    }

    private _onMessage(event: MessageEvent): void {
        const message = JSON.parse(event.data) as ChatService.ServerMessage;
        if (\\"operation\\" in message) {
            console.log(\\"Received server message\\");
        }
        else {
            const callback = this.callbacks[message.replyTo];
            if (callback != null) {
                callback(message);
            }
            else {
                console.error(\\"Received reply to unknown request\\");
            }
        }
    }
}
",
  "/src/services/chat/index.ts": "export * from \\"./ChatService\\";
",
  "/tsconfig.esm.json": "{
    \\"compilerOptions\\": {
        \\"composite\\": true,
        \\"module\\": \\"esnext\\",
        \\"target\\": \\"esnext\\",
        \\"outDir\\": \\"lib/esm\\",
        \\"rootDir\\": \\"src\\",
        \\"moduleResolution\\": \\"node\\",
        \\"esModuleInterop\\": true,
        \\"strict\\": true,
        \\"declaration\\": true,
        \\"noFallthroughCasesInSwitch\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"noUncheckedIndexedAccess\\": true,
        \\"noUnusedLocals\\": true,
        \\"noUnusedParameters\\": true
    },
    \\"include\\": [
        \\"src\\"
    ]
}",
  "/tsconfig.json": "{
    \\"compilerOptions\\": {
        \\"composite\\": true,
        \\"module\\": \\"commonjs\\",
        \\"target\\": \\"esnext\\",
        \\"outDir\\": \\"lib/cjs\\",
        \\"rootDir\\": \\"src\\",
        \\"moduleResolution\\": \\"node\\",
        \\"esModuleInterop\\": true,
        \\"strict\\": true,
        \\"declaration\\": true,
        \\"noFallthroughCasesInSwitch\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"noUncheckedIndexedAccess\\": true,
        \\"noUnusedLocals\\": true,
        \\"noUnusedParameters\\": true
    },
    \\"include\\": [
        \\"src\\"
    ]
}",
}
`;

exports[`generateClientProject no errors 1`] = `
Object {
  "/package.json": "{
    \\"name\\": \\"no-errors\\",
    \\"version\\": \\"0.0.0\\",
    \\"files\\": [
        \\"lib\\"
    ],
    \\"main\\": \\"./lib/cjs/index.js\\",
    \\"types\\": \\"./lib/cjs/index.d.ts\\",
    \\"exports\\": {
        \\".\\": {
            \\"require\\": \\"./lib/cjs/index.js\\",
            \\"default\\": \\"./lib/esm/index.js\\"
        }
    },
    \\"sideEffects\\": false,
    \\"scripts\\": {
        \\"build\\": \\"run-p build:esm build:cjs\\",
        \\"build:esm\\": \\"tsc --project tsconfig.esm.json && echo '{ \\\\\\"type\\\\\\": \\\\\\"module\\\\\\" }' > lib/esm/package.json\\",
        \\"build:cjs\\": \\"tsc --project tsconfig.json && echo '{ \\\\\\"type\\\\\\": \\\\\\"commonjs\\\\\\" }' > lib/cjs/package.json\\"
    },
    \\"dependencies\\": {
        \\"@fern-typescript/service-utils\\": \\"0.0.86\\"
    },
    \\"devDependencies\\": {
        \\"@types/node\\": \\"^17.0.33\\",
        \\"npm-run-all\\": \\"^4.1.5\\",
        \\"typescript\\": \\"^4.6.4\\"
    }
}",
  "/src/model/api/index.ts": "export * from \\"./service-types\\";
",
  "/src/model/api/service-types/NoErrorsService/_GetResponse.ts": "export type _GetResponse = _GetResponse.SuccessResponse | _GetResponse.ErrorResponse;

namespace _GetResponse {
    export interface SuccessResponse {
        ok: true;
        statusCode: number;
    }

    export interface ErrorResponse {
        ok: false;
        statusCode: number;
    }
}
",
  "/src/model/api/service-types/NoErrorsService/index.ts": "export * from \\"./_GetResponse\\";
",
  "/src/model/api/service-types/index.ts": "export * from \\"./NoErrorsService\\";
",
  "/src/model/index.ts": "export * as api from \\"./api\\";
",
  "/src/services/api/NoErrorsService.ts": "import { Fetcher, defaultFetcher, Service, isResponseOk, Token } from \\"@fern-typescript/service-utils\\";
import * as model from \\"../../model\\";
import path from \\"path\\";

export interface NoErrorsService {
    get(): Promise<model.api._GetResponse>;
}

export class NoErrorsService implements NoErrorsService {
    private baseUrl: string;
    private fetcher: Fetcher;
    private token: Token | undefined;

    constructor(args: Service.Init) {
        this.fetcher = args.fetcher ?? defaultFetcher;
        this.baseUrl = args.origin;
        this.token = args.token;
    }

    public async get(): Promise<model.api._GetResponse> {
        const encodedResponse = await this.fetcher({
            url: path.join(this.baseUrl, \\"/\\"),
            method: \\"GET\\",
            headers: {},
            token: this.token
        });

        if (isResponseOk(encodedResponse)) {
            return {
                ok: true,
                statusCode: encodedResponse.statusCode
            };
        }
        else {
            return {
                ok: false,
                statusCode: encodedResponse.statusCode
            };
        }
    }
}
",
  "/src/services/api/index.ts": "export * from \\"./NoErrorsService\\";
",
  "/tsconfig.esm.json": "{
    \\"compilerOptions\\": {
        \\"composite\\": true,
        \\"module\\": \\"esnext\\",
        \\"target\\": \\"esnext\\",
        \\"outDir\\": \\"lib/esm\\",
        \\"rootDir\\": \\"src\\",
        \\"moduleResolution\\": \\"node\\",
        \\"esModuleInterop\\": true,
        \\"strict\\": true,
        \\"declaration\\": true,
        \\"noFallthroughCasesInSwitch\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"noUncheckedIndexedAccess\\": true,
        \\"noUnusedLocals\\": true,
        \\"noUnusedParameters\\": true
    },
    \\"include\\": [
        \\"src\\"
    ]
}",
  "/tsconfig.json": "{
    \\"compilerOptions\\": {
        \\"composite\\": true,
        \\"module\\": \\"commonjs\\",
        \\"target\\": \\"esnext\\",
        \\"outDir\\": \\"lib/cjs\\",
        \\"rootDir\\": \\"src\\",
        \\"moduleResolution\\": \\"node\\",
        \\"esModuleInterop\\": true,
        \\"strict\\": true,
        \\"declaration\\": true,
        \\"noFallthroughCasesInSwitch\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"noUncheckedIndexedAccess\\": true,
        \\"noUnusedLocals\\": true,
        \\"noUnusedParameters\\": true
    },
    \\"include\\": [
        \\"src\\"
    ]
}",
}
`;

exports[`generateClientProject posts 1`] = `
Object {
  "/package.json": "{
    \\"name\\": \\"posts\\",
    \\"version\\": \\"0.0.0\\",
    \\"files\\": [
        \\"lib\\"
    ],
    \\"main\\": \\"./lib/cjs/index.js\\",
    \\"types\\": \\"./lib/cjs/index.d.ts\\",
    \\"exports\\": {
        \\".\\": {
            \\"require\\": \\"./lib/cjs/index.js\\",
            \\"default\\": \\"./lib/esm/index.js\\"
        }
    },
    \\"sideEffects\\": false,
    \\"scripts\\": {
        \\"build\\": \\"run-p build:esm build:cjs\\",
        \\"build:esm\\": \\"tsc --project tsconfig.esm.json && echo '{ \\\\\\"type\\\\\\": \\\\\\"module\\\\\\" }' > lib/esm/package.json\\",
        \\"build:cjs\\": \\"tsc --project tsconfig.json && echo '{ \\\\\\"type\\\\\\": \\\\\\"commonjs\\\\\\" }' > lib/cjs/package.json\\"
    },
    \\"dependencies\\": {
        \\"uuid\\": \\"^8.3.2\\",
        \\"@fern-typescript/service-utils\\": \\"0.0.86\\"
    },
    \\"devDependencies\\": {
        \\"@types/uuid\\": \\"^8.3.4\\",
        \\"@types/node\\": \\"^17.0.33\\",
        \\"npm-run-all\\": \\"^4.1.5\\",
        \\"typescript\\": \\"^4.6.4\\"
    }
}",
  "/src/model/commons/index.ts": "export * from \\"./types\\";
",
  "/src/model/commons/types/PostId.ts": "/**
 * An ID of a blog post
 */
export type PostId = string & {
    __PostId: void;
};

export const PostId = {
    of: (value: string): PostId => value as PostId
};
",
  "/src/model/commons/types/index.ts": "export * from \\"./PostId\\";
",
  "/src/model/index.ts": "export * as commons from \\"./commons\\";
export * as posts from \\"./posts\\";
",
  "/src/model/posts/errors/PostNotFoundError.ts": "import * as model from \\"../..\\";

export interface PostNotFoundError {
    id: model.commons.PostId;
}
",
  "/src/model/posts/errors/index.ts": "export * from \\"./PostNotFoundError\\";
",
  "/src/model/posts/index.ts": "export * from \\"./types\\";
export * from \\"./errors\\";
export * from \\"./service-types\\";
",
  "/src/model/posts/service-types/PostsService/_CreatePostRequest.ts": "import * as model from \\"../../..\\";

export interface _CreatePostRequest {
    type: model.posts.PostType;
    title: string;
    author: model.posts.Author;
    content: string;
}
",
  "/src/model/posts/service-types/PostsService/_CreatePostResponse.ts": "import * as model from \\"../../..\\";

export type _CreatePostResponse = _CreatePostResponse.SuccessResponse | _CreatePostResponse.ErrorResponse;

namespace _CreatePostResponse {
    export interface SuccessResponse {
        ok: true;
        statusCode: number;
        body: model.commons.PostId;
    }

    export interface ErrorResponse {
        ok: false;
        statusCode: number;
    }
}
",
  "/src/model/posts/service-types/PostsService/_GetPostErrorBody.ts": "import { posts } from \\"../../..\\";
import * as uuid from \\"uuid\\";

export type _GetPostErrorBody =
    | _GetPostErrorBody.PostNotFoundError;

export declare namespace _GetPostErrorBody {
    interface PostNotFoundError extends posts.PostNotFoundError {
        _error: \\"PostNotFoundError\\";
        _errorInstanceId: string;
    }

    export interface _Visitor<Result> {
        PostNotFoundError: (value: posts.PostNotFoundError) => Result;
        _unknown: () => Result;
    }
}

export const _GetPostErrorBody = {
    PostNotFoundError: (value: posts.PostNotFoundError): _GetPostErrorBody.PostNotFoundError => ({
        ...value,
        _error: \\"PostNotFoundError\\",
        _errorInstanceId: uuid.v4()
    }),

    _visit: <Result>(value: _GetPostErrorBody, visitor: _GetPostErrorBody._Visitor<Result>): Result => {
        switch (value._error) {
            case \\"PostNotFoundError\\": return visitor.PostNotFoundError(value);
            default: return visitor._unknown();
        }
    },

    _types: (): _GetPostErrorBody[\\"_error\\"][] => [\\"PostNotFoundError\\"],
} as const;
",
  "/src/model/posts/service-types/PostsService/_GetPostRequest.ts": "import * as model from \\"../../..\\";

export interface _GetPostRequest {
    postId: model.commons.PostId;
    page: number | null | undefined;
    otherParam: number | null | undefined;
}
",
  "/src/model/posts/service-types/PostsService/_GetPostResponse.ts": "import * as model from \\"../../..\\";

export type _GetPostResponse = _GetPostResponse.SuccessResponse | _GetPostResponse.ErrorResponse;

namespace _GetPostResponse {
    export interface SuccessResponse {
        ok: true;
        statusCode: number;
        body: model.posts.Post;
    }

    export interface ErrorResponse {
        ok: false;
        statusCode: number;
        error: model.posts._GetPostErrorBody;
    }
}
",
  "/src/model/posts/service-types/PostsService/_GetPostV2ErrorBody.ts": "import { posts } from \\"../../..\\";
import * as uuid from \\"uuid\\";

export type _GetPostV2ErrorBody =
    | _GetPostV2ErrorBody.PostNotFoundError;

export declare namespace _GetPostV2ErrorBody {
    interface PostNotFoundError extends posts.PostNotFoundError {
        _error: \\"PostNotFoundError\\";
        _errorInstanceId: string;
    }

    export interface _Visitor<Result> {
        PostNotFoundError: (value: posts.PostNotFoundError) => Result;
        _unknown: () => Result;
    }
}

export const _GetPostV2ErrorBody = {
    PostNotFoundError: (value: posts.PostNotFoundError): _GetPostV2ErrorBody.PostNotFoundError => ({
        ...value,
        _error: \\"PostNotFoundError\\",
        _errorInstanceId: uuid.v4()
    }),

    _visit: <Result>(value: _GetPostV2ErrorBody, visitor: _GetPostV2ErrorBody._Visitor<Result>): Result => {
        switch (value._error) {
            case \\"PostNotFoundError\\": return visitor.PostNotFoundError(value);
            default: return visitor._unknown();
        }
    },

    _types: (): _GetPostV2ErrorBody[\\"_error\\"][] => [\\"PostNotFoundError\\"],
} as const;
",
  "/src/model/posts/service-types/PostsService/_GetPostV2Response.ts": "import * as model from \\"../../..\\";

export type _GetPostV2Response = _GetPostV2Response.SuccessResponse | _GetPostV2Response.ErrorResponse;

namespace _GetPostV2Response {
    export interface SuccessResponse {
        ok: true;
        statusCode: number;
        body: model.posts.Post;
    }

    export interface ErrorResponse {
        ok: false;
        statusCode: number;
        error: model.posts._GetPostV2ErrorBody;
    }
}
",
  "/src/model/posts/service-types/PostsService/index.ts": "export * from \\"./_CreatePostRequest\\";
export * from \\"./_CreatePostResponse\\";
export * from \\"./_GetPostRequest\\";
export * from \\"./_GetPostErrorBody\\";
export * from \\"./_GetPostResponse\\";
export * from \\"./_GetPostV2ErrorBody\\";
export * from \\"./_GetPostV2Response\\";
",
  "/src/model/posts/service-types/index.ts": "export * from \\"./PostsService\\";
",
  "/src/model/posts/types/Author.ts": "export type Author =
    | Author.Anonymous
    | Author.Named;

export declare namespace Author {
    interface Anonymous {
        _type: \\"anonymous\\";
    }

    interface Named {
        _type: \\"named\\";
        named: string;
    }

    export interface _Visitor<Result> {
        anonymous: () => Result;
        named: (value: string) => Result;
        _unknown: () => Result;
    }
}

export const Author = {
    anonymous: (): Author.Anonymous => ({
        _type: \\"anonymous\\"
    }),

    named: (value: string): Author.Named => ({
        named: value,
        _type: \\"named\\"
    }),

    _visit: <Result>(value: Author, visitor: Author._Visitor<Result>): Result => {
        switch (value._type) {
            case \\"anonymous\\": return visitor.anonymous();
            case \\"named\\": return visitor.named(value.named);
            default: return visitor._unknown();
        }
    },

    _types: (): Author[\\"_type\\"][] => [\\"anonymous\\", \\"named\\"],
} as const;
",
  "/src/model/posts/types/Post.ts": "import * as model from \\"../..\\";

/**
 * A blog post
 */
export interface Post {
    id: model.commons.PostId;
    type: model.posts.PostType;
    title: string;
    author: model.posts.Author;
    content: string;
    relatedPosts: model.posts.Post[];
}
",
  "/src/model/posts/types/PostType.ts": "export type PostType =
    | \\"LONG\\"
    | \\"SHORT\\";

export const PostType = {
    Long: \\"LONG\\",
    Short: \\"SHORT\\",

    _visit: <Result>(value: PostType, visitor: PostType._Visitor<Result>): Result => {
        switch (value) {
            case PostType.Long: return visitor.long();
            case PostType.Short: return visitor.short();
            default: return visitor._unknown();
        }
    },

    _values: (): PostType[] => [PostType.Long, PostType.Short],
} as const;

export declare namespace PostType {
    type Long = \\"LONG\\";
    type Short = \\"SHORT\\";

    export interface _Visitor<Result> {
        long: () => Result;
        short: () => Result;
        _unknown: () => Result;
    }
}
",
  "/src/model/posts/types/index.ts": "export * from \\"./Post\\";
export * from \\"./PostType\\";
export * from \\"./Author\\";
",
  "/src/services/posts/PostsService.ts": "import { Fetcher, defaultFetcher, Service, isResponseOk, Token } from \\"@fern-typescript/service-utils\\";
import path from \\"path\\";
import * as model from \\"../../model\\";

export interface PostsService {
    createPost(request: model.posts._CreatePostRequest): Promise<model.posts._CreatePostResponse>;
    getPost(request: model.posts._GetPostRequest): Promise<model.posts._GetPostResponse>;
    getPostV2(request: model.commons.PostId): Promise<model.posts._GetPostV2Response>;
}

export class PostsService implements PostsService {
    private baseUrl: string;
    private fetcher: Fetcher;
    private token: Token | undefined;

    constructor(args: Service.Init) {
        this.fetcher = args.fetcher ?? defaultFetcher;
        this.baseUrl = path.join(args.origin, \\"/posts\\");
        this.token = args.token;
    }

    public async createPost(request: model.posts._CreatePostRequest): Promise<model.posts._CreatePostResponse> {
        const encodedResponse = await this.fetcher({
            url: path.join(this.baseUrl, \\"/\\"),
            method: \\"POST\\",
            headers: {},
            token: this.token,
            body: {
                content: JSON.stringify(request),
                contentType: \\"application/json\\"
            }
        });

        if (isResponseOk(encodedResponse)) {
            const response = JSON.parse(new TextDecoder().decode(encodedResponse.body));
            return {
                ok: true,
                statusCode: encodedResponse.statusCode,
                body: response as model.commons.PostId
            };
        }
        else {
            return {
                ok: false,
                statusCode: encodedResponse.statusCode
            };
        }
    }

    public async getPost(request: model.posts._GetPostRequest): Promise<model.posts._GetPostResponse> {
        const queryParameters = new URLSearchParams();
        if (request.page != null) {
            queryParameters.append(\\"page\\", request.page.toString());
        }
        if (request.otherParam != null) {
            queryParameters.append(\\"otherParam\\", request.otherParam.toString());
        }

        const encodedResponse = await this.fetcher({
            url: path.join(this.baseUrl, \`/\${request.postId}\`),
            method: \\"GET\\",
            headers: {},
            token: this.token,
            queryParameters
        });

        if (isResponseOk(encodedResponse)) {
            const response = JSON.parse(new TextDecoder().decode(encodedResponse.body));
            return {
                ok: true,
                statusCode: encodedResponse.statusCode,
                body: response as model.posts.Post
            };
        }
        else {
            const error = JSON.parse(new TextDecoder().decode(encodedResponse.body));
            return {
                ok: false,
                statusCode: encodedResponse.statusCode,
                error: error as model.posts._GetPostErrorBody
            };
        }
    }

    public async getPostV2(request: model.commons.PostId): Promise<model.posts._GetPostV2Response> {
        const encodedResponse = await this.fetcher({
            url: path.join(this.baseUrl, \\"/get\\"),
            method: \\"GET\\",
            headers: {},
            token: this.token,
            body: {
                content: JSON.stringify(request),
                contentType: \\"application/json\\"
            }
        });

        if (isResponseOk(encodedResponse)) {
            const response = JSON.parse(new TextDecoder().decode(encodedResponse.body));
            return {
                ok: true,
                statusCode: encodedResponse.statusCode,
                body: response as model.posts.Post
            };
        }
        else {
            const error = JSON.parse(new TextDecoder().decode(encodedResponse.body));
            return {
                ok: false,
                statusCode: encodedResponse.statusCode,
                error: error as model.posts._GetPostV2ErrorBody
            };
        }
    }
}
",
  "/src/services/posts/index.ts": "export * from \\"./PostsService\\";
",
  "/tsconfig.esm.json": "{
    \\"compilerOptions\\": {
        \\"composite\\": true,
        \\"module\\": \\"esnext\\",
        \\"target\\": \\"esnext\\",
        \\"outDir\\": \\"lib/esm\\",
        \\"rootDir\\": \\"src\\",
        \\"moduleResolution\\": \\"node\\",
        \\"esModuleInterop\\": true,
        \\"strict\\": true,
        \\"declaration\\": true,
        \\"noFallthroughCasesInSwitch\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"noUncheckedIndexedAccess\\": true,
        \\"noUnusedLocals\\": true,
        \\"noUnusedParameters\\": true
    },
    \\"include\\": [
        \\"src\\"
    ]
}",
  "/tsconfig.json": "{
    \\"compilerOptions\\": {
        \\"composite\\": true,
        \\"module\\": \\"commonjs\\",
        \\"target\\": \\"esnext\\",
        \\"outDir\\": \\"lib/cjs\\",
        \\"rootDir\\": \\"src\\",
        \\"moduleResolution\\": \\"node\\",
        \\"esModuleInterop\\": true,
        \\"strict\\": true,
        \\"declaration\\": true,
        \\"noFallthroughCasesInSwitch\\": true,
        \\"forceConsistentCasingInFileNames\\": true,
        \\"noUncheckedIndexedAccess\\": true,
        \\"noUnusedLocals\\": true,
        \\"noUnusedParameters\\": true
    },
    \\"include\\": [
        \\"src\\"
    ]
}",
}
`;
